### ISSUE TITLE
itest flake: subsystem height de-sync issue "update announcement is premature"

### ISSUE THREAD

**Issue by guggero** [lightningnetwork/lnd#5903]
Seeing these kinds of failures from time to time in itests.

The underlying cause is always the `DISC: Update announcement for short_chan_id(803742999969792), is premature: advertises height 731, only height 730 is known` error.
The discovery subsystem seems to be on a different best known height as the peer subsystem which leads to a channel announcement being ignored and subsequently the channel not being found in the node's graph.

I think we might need a **synchronous** way of processing updates to the best known height. A PoC with my proposed solution on how to fix this is here: https://github.com/guggero/lnd/tree/sync-subsystem-block-height

Log in GitHub action:
```
     --- FAIL: TestLightningNetworkDaemon/tranche01/29-of-89/neutrino/invoice_routing_hints (48.93s)
     test_harness.go:90: Failed: (invoice routing hints): exited with error: 
        *errors.errorString timed out waiting for channel open alice-bob: channel:35345a3671c3822ebd3824b1796919b0328eace57d6f5616a4c3d534b4cceb8c:0 not opened before timeout: 
        node state: {
        	"NodeID": 34,
        	"Name": "Alice",
        	"PubKey": "02e5fe031b08ffd687af72cfee39baca946545bfbdf463c131946b45d19a1e2439",
        	"OpenChans": {
        		"104bacf664203b730a09da38288c17e37d15bf08220e102ffe00668e166503f4:0": 2,
        		"35345a3671c3822ebd3824b1796919b0328eace57d6f5616a4c3d534b4cceb8c:0": 1,
        		"76b773fe8ba9c16b53f683ad85d2be4678d015671a273ad05cd4f7687fea0a87:0": 2,
        		"dc0c786ebd8d9f7e14dad5b35316e2e707f465e8cf168d0b8265970e8c8c919f:0": 2,
        		"dd0ce8c3c7596607649c041be0fd5c5530bfef8a6163dcda5efd78d19696babe:0": 2
        	},
        	"ClosedChans": {},
        	"NodeCfg": {
        		"LogFilenamePrefix": "invoice_routing_hints",
        		"ExtraArgs": [
        			"--default-remote-max-htlcs=483",
        			"--dust-threshold=5000000"
        		],
        		"HasSeed": false,
        		"P2PPort": 6692,
        		"RPCPort": 6696,
        		"RESTPort": 6697,
        		"ProfilePort": 0,
        		"AcceptKeySend": true,
        		"AcceptAMP": false,
        		"FeeURL": "http://localhost:6558/fee-estimates.json"
        	}
        }
        /home/runner/work/lnd/lnd/lntest/itest/lnd_routing_test.go:1148 (0xf28ff7)
        	testInvoiceRoutingHints: t.Fatalf("timed out waiting for channel open %s: %v",
        /home/runner/work/lnd/lnd/lntest/itest/test_harness.go:114 (0xeb84dc)
        	(*harnessTest).RunTestCase: testCase.test(h.lndHarness, h)
        /home/runner/work/lnd/lnd/lntest/itest/lnd_test.go:250 (0xf3bcda)
        	TestLightningNetworkDaemon.func4: ht.RunTestCase(testCase)
        /opt/hostedtoolcache/go/1.17.1/x64/src/testing/testing.go:1259 (0x509f62)
        	tRunner: fn(t)
        /opt/hostedtoolcache/go/1.17.1/x64/src/runtime/asm_amd64.s:1581 (0x468661)
        	goexit: BYTE	$0x90	// NOP
```

Here's the relevant entry from the log (Alice, see `34-invoice_routing_hints-Alice-02e5fe03.log`):
```
2021-10-29 00:28:47.263 [DBG] PEER: Received ChannelUpdate(chain_hash=0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206, short_chan_id=803742999969792, mflags=00000001, cflags=00000001, update_time=2021-10-29 00:28:47 +0000 UTC) from 03a11721372ebd31b60fe7c6e802714a37d09de5f979945711428464bce095d03b@127.0.0.1:6691
2021-10-29 00:28:47.263 [INF] DISC: Update announcement for short_chan_id(803742999969792), is premature: advertises height 731, only height 730 is known
```

And here are the full logs from that run:
[itest-neutrino-logs.zip](https://github.com/lightningnetwork/lnd/files/7439986/itest-neutrino-logs.zip)

**Comment by vertiond** [comment]
I just experienced a similar issue which brought my node offline for a couple days.  It would hang upon trying to shutdown.

Bits I thought might be relevant:
```
2022-01-08 20:51:36.459 [INF] CRTR: Pruning channel graph using block 000000000000000000052b3558670391271e50a79ceeac2bb7d69e8cee2e44af (height=717777)
2022-01-08 20:51:36.587 [INF] CRTR: Block 000000000000000000052b3558670391271e50a79ceeac2bb7d69e8cee2e44af (height=717777) closed 0 channels
2022-01-08 20:51:36.603 [INF] NTFN: New block: height=717777, sha=000000000000000000052b3558670391271e50a79ceeac2bb7d69e8cee2e44af
2022-01-08 20:51:36.635 [INF] UTXN: Attempting to graduate height=717777: num_kids=0, num_babies=0
2022-01-08 20:51:36.635 [INF] CRTR: Pruning channel graph using block 000000000000000000052b3558670391271e50a79ceeac2bb7d69e8cee2e44af (height=717777)
2022-01-08 20:51:36.641 [INF] NTFN: Block disconnected from main chain: height=717777, sha=000000000000000000052b3558670391271e50a79ceeac2bb7d69e8cee2e44af
2022-01-08 20:51:36.648 [INF] HSWC: Garbage collected 8 shared secret hashes at height=717777
2022-01-08 20:51:36.735 [INF] CRTR: Block 000000000000000000052b3558670391271e50a79ceeac2bb7d69e8cee2e44af (height=717777) closed 0 channels
2022-01-08 20:51:36.770 [INF] NTFN: New block: height=717777, sha=000000000000000000052b3558670391271e50a79ceeac2bb7d69e8cee2e44af
2022-01-08 20:51:36.773 [INF] UTXN: Attempting to graduate height=717777: num_kids=0, num_babies=0

2022-01-08 20:55:48.651 [INF] CRTR: Pruning channel graph using block 000000000000000000037300573eeb5a6618daa8dffb075e0fd4097034223c76 (height=717778)
2022-01-08 20:55:48.710 [INF] DISC: Received new remote channel announcement for 717773:1589:1
2022-01-08 20:55:48.806 [INF] DISC: 1/2 of channel ann proof received for short_chan_id=789199759707799553, waiting for other half
2022-01-08 20:55:48.808 [INF] CRTR: Block 000000000000000000037300573eeb5a6618daa8dffb075e0fd4097034223c76 (height=717778) closed 0 channels
2022-01-08 20:55:49.052 [INF] NTFN: New block: height=717778, sha=000000000000000000037300573eeb5a6618daa8dffb075e0fd4097034223c76
2022-01-08 20:55:49.062 [INF] UTXN: Attempting to graduate height=717778: num_kids=0, num_babies=0
2022-01-08 20:55:49.068 [ERR] LTND: Unable to lookup witness: no witnesses
2022-01-08 20:55:49.101 [INF] CRTR: Pruning channel graph using block 000000000000000000037300573eeb5a6618daa8dffb075e0fd4097034223c76 (height=717778)
2022-01-08 20:55:49.102 [INF] HSWC: Garbage collected 8 shared secret hashes at height=717778
2022-01-08 20:55:49.199 [INF] CRTR: Block 000000000000000000037300573eeb5a6618daa8dffb075e0fd4097034223c76 (height=717778) closed 0 channels

2022-01-08 22:09:15.097 [INF] DISC: GossipSyncer(0347ad98277e439d8027e2922fd2e37e016339a622908a30ccb1fb4fdca87cadfc): requesting new chans from height=0 and 717778 blocks after

2022-01-08 22:32:03.711 [INF] DISC: Announcement for chan
_id=(789207456238338052), is premature: advertises height 717780, only height 717778 is known

2022-01-09 20:52:30.565 [INF] DISC: Announcement for chan_id=(789350392818630657), is premature: advertises height 717910, only height 717778 is known

2022-01-11 09:00:04.270 [ERR] RPCS: [/lnrpc.Lightning/GetChanInfo]: edge not found

2022-01-12 07:09:15.097 [INF] DISC: GossipSyncer(037b6d303c95b4faf2f62a214cc32c78aa0ded8ab5bd7a11aaa4883bbe292a4764): requesting new chans from height=0 and 717778 blocks after

```

**Comment by yyforyongyu** [comment]
> I just experienced a similar issue which brought my node offline for a couple days. It would hang upon trying to shutdown.

Which `lnd` version are you running?

**Comment by vertiond** [comment]
lnd v0.14.1-beta @yyforyongyu

**Comment by yyforyongyu** [comment]
> Announcement for chan_id=(789207456238338052), is premature: advertises height 717780, only height 717778 is known

This could happen when your node is left behind with some blocks, and it's fixed in #6054. However this shouldn't cause your node to shut down. Do you have more logs to show that your node was offline?

**Comment by vertiond** [comment]
Sorry, I should clarify - my node did not shut down but most of my channels were disabled.  I did also have an issue where `lncli stop` was failing to fully stop lnd and ended up rebooting.

**Comment by yyforyongyu** [comment]
fixed by #6825
